### PR TITLE
[COMPASS-1851] - Optimise the build of Compass Plugin template

### PR DESCRIPTION
**Pull request by matt-d-rat**
# Summary

This PR significantly reduces the output size of the compiled library bundle of the plugin, by externalising more dependencies as peers, tweaking the minification/transformation settings. As you can see from the screenshots below the build times have gone from `~16.6 seconds (324kB)` to just `~2.1 seconds (19.6kB)`.

## Before Optimisations

compass-plugin-stats-before

compass-plugin-analysis-before

## After Optimisations

compass-plugin-stats-after-01

compass-plugin-analysis-after-01

# Changes

- Reduced the output size of the bundle file by declaring common Compass dependencies as peerDependencies on the plugin.
- Updated webpack configurations to fix linting errors.
- Removed prod `electron.js` bundle output from webpack configuration as this was causing builds to indefinitely hang on larger plugin code bases.
- Added webpack bundle analyzer as a devDependency.

### CODE REVIEW

File: template/config/webpack.test.config.js [hunk 1]
```diff
@@ -82,20 +72,20 @@ module.exports = {
...
-              /.*?(?=\.spec).*?\.js/
-          ],
-          include: project.path.src,
-          use: {
-              loader: 'istanbul-instrumenter-loader',
-              options: {
-                  esModules: true
-              }
+        test: /\.(js|jsx)/,
+        enforce: 'post', // Enforce as a post step so babel can do its compliation prior to instrumenting code
```
Comment (pzrq) [inline template/config/webpack.test.config.js]: compliation  -> compilation?

Comment (matt-d-rat) [reply template/config/webpack.test.config.js]: loaders are not necessarily for compilation, although that's what they are normally implemented to do. A good analogy is to think of the stages in the MongoDB aggregation pipeline, where loaders would be the stages, except they may be configured to work on different data. In this case, the instrument-loader simply takes the compiled output and uses it in its stage to run the nyc cli.

Comment (matt-d-rat) [reply template/config/webpack.test.config.js]: Sorry complete misunderstood what you meant. I think I might be dyslexic as I couldn't see that that word had a typo. I have updated the comment correcting the mistake.

File: template/package.json [hunk 2]
```diff
@@ -7,17 +7,17 @@
...
     "start:watch": "npm run clean && webpack --config ./config/webpack.watch.config.js",
-    "start:prod": "npm run compile && electron --noDevServer ./electron",
     "test": "cross-env NODE_ENV=test mocha-webpack \"./src/**/*.spec.js\"",
     "test:watch": "cross-env NODE_ENV=test mocha-webpack \"./src/**/*.spec.js\" --watch",
     "test:karma": "cross-env NODE_ENV=test karma start",
     "cover": "nyc npm run test",
     "ci": "npm run check && npm test",
     "fmt": "mongodb-js-fmt ./*.js ./test/*.js",
-    "check": "mongodb-js-precommit './src/**/*{.js,.jsx}' './test/**/*.js' index.js",
+    "precheck": "npm run compile",
```
Comment (rueckstiess) [inline template/package.json]: is compilation necessary as a precheck step if the check itself only runs in ./src and ./test?

File: template/config/webpack.prod.config.js [hunk 3]
```diff
@@ -177,9 +134,12 @@ module.exports = {
     // Creates HTML page for us at build time
     new HtmlWebpackPlugin(),
 
-    // An ES6+ aware minifier, results in smaller output compared to UglifyJS given that 
+    // An ES6+ aware minifier, results in smaller output compared to UglifyJS given that
     // Chromium in electron supports the majority of ES6 features out of the box.
-    new MinifyPlugin()
+    new MinifyPlugin(),
```
Comment (rueckstiess) [inline template/config/webpack.prod.config.js]: why did you add the comma here? Isn't it the last element in the array?

File: template/.babelrc [hunk 4]
```diff
@@ -1,7 +1,8 @@
 {
   "presets": [
   	"react",
-  	['es2015', {'modules': false}]
+  	['es2015', {'modules': false}],
+    "minify"
```
Comment (rueckstiess) [inline template/.babelrc]: The indentation looks funky here for these lines.

File: template/config/webpack.prod.config.js (hunk 3 above)
Comment (rueckstiess) [reply template/config/webpack.prod.config.js]: Ah, got it, because of the commented out BundleAnalyzerPlugin. nevermind.

File: template/.babelrc (hunk 4 above)
Comment (matt-d-rat) [reply template/.babelrc]: - [ ] Fix indentation

File: template/package.json (hunk 2 above)
Comment (rueckstiess) [reply template/package.json]: as discussed. needs to be done because check also goes through the package.json main entry point.

File: template/config/webpack.prod.config.js (hunk 3 above)
Comment (matt-d-rat) [reply template/config/webpack.prod.config.js]: So that you can simply uncomment the line below it if you want to run the analyser over the build.

@rueckstiess I can remove the dangling comma if you prefer?

**Comment by matt-d-rat** [comment]
Ignore the revert branch. It accidentally got created when I over eagerly pressed the merge button which turns into a revert button in the UI. Silly UX. I have deleted that remote branch. No revert commits were merged.
